flang/test/Semantics/OpenMP/target01.f90

```diff
@@ -39,7 +39,6 @@ subroutine bar(b1, b2, b3)
   type(c_ptr), pointer :: b2
   type(c_ptr), value :: b3
 
-  !ERROR: Variable 'c' in IS_DEVICE_PTR clause must be a dummy argument
```

Comment:
According to the restriction, this error should be reported right? It says argument must be a dummy argument. Am I missing something?

Reply:
I think this restriction is something on the lines of "If the list-item in an is_device_ptr clause is a dummy argument it should not have ALLOCATABLE, POINTER or VALUE attributes."

I think this following text might not have completely meant that all list-items must be dummy arguments only I think the conjunction of "does not have the ALLOCATABLE, POINTER or VALUE attribute" might be the reason for the "must". 
`A list item that appears in an is_device_ptr clause must be a dummy argument that does not have the ALLOCATABLE, POINTER or VALUE attribute.`

Also, this restriction is not present from OpenMP 5.1 onwards through which I assumed it might not be originally intended for OpenMP spec committee might not have intended to restricted list-item of is_device_ptr clause to be dummy argument only.

Please correct me if I am interpreting it incorrectly. :)

Reply:
So it seems like the type of list item became more restrictive as the standard progressed.

OpenMP 4.5 standard (2.10.4): `A list item that appears in an is_device_ptr clause must be a dummy argument`

OpenMP 5.0 standard (2.12.5): `A list item that appears in an is_device_ptr clause must be a dummy argument that does not have the ALLOCATABLE, POINTER or VALUE attribute.`

However, my understanding of both these statements is leaning similar to @shraiysh's- any list item in this clause has to be a dummy argument, but without any additional attributes from ALLOCATABLE, POINTER, VALUE. In other words, I don't think non-dummy arguments are allowed in this clause. This becomes clear when we look at the OpenMP 4.5 standard wording: it is, without a doubt, restricting all list items within the clause to be dummy arguments.

Reply:
I cannot find this check, or anything related to the c_ptr being a dummy argument in the OpenMP 5.2 spec. If this check is not required in 5.2, we probably should remove this check completely - please let me know if there is anything related to this in 5.2. As discussed in the meeting, OpenMP 5.2 is essentially a fix on OpenMP 5.1 and OpenMP 5.0, so we should stick to OpenMP 5.2.

Reply:
I think in 5.2 we deprecate all the other use of the **is_device_ptr** clause in Fortran except `type(c_ptr)` variables in order to avoid any confusion between pointer and nonpointer list item. The **has_device_address** clause should be the choice for Fortran unless it is a `type(c_ptr)` variable.

In Section 5.4.7, we have:
> If the **is_device_ptr** clause is specified on a **target** construct, if any list item is not of type **C_PTR**, the behavior is as if the list item appeared in a **has_device_addr** clause. Support for such list items in an **is_device_ptr** clause is deprecated.

In my opinion, issuing a warning message about the use being deprecated is good for the users. Furthermore, I don't think the spec wants to add more Fortran stuff to the **is_device_ptr** clause as it creates so much confusion in the past.